### PR TITLE
Federation Setup Console.

### LINKED ISSUE #190: Federation Set Up Console
Add `GenesisMiner` and `MultisigAddressCreator` to the console app.

### PR DESCRIPTION
**Pull request by ferdeen**
@bokobza, I've added GenesisMiner and MultisigAddressCreator to console app.

image

@monsieurleberre, do we need to include the creation of public private keys (-p) - I kept the code in there as wasn't too sure.  

Fixes #190.

Cheers.

### CODE REVIEW

File: src/FedKeyPairGen/Program.cs [hunk 1]
```diff
@@ -27,37 +30,53 @@ static void Main(string[] args)
...
+                {
+                    Console.WriteLine(new MultisigAddressCreator().CreateMultisigAddresses(
+                        Networks.Stratis.Testnet(),
+                        FederatedPegNetwork.NetworksSelector.Testnet()));
+                }
+
+                if (args.Contains("-p"))
+                {
+                    var mnemonic = new Mnemonic(Wordlist.English, WordCount.Twelve);
+                    var pubKey = mnemonic.DeriveExtKey().PrivateKey.PubKey;
```
Comment (ferdeen) [inline src/FedKeyPairGen/Program.cs]: @monsieurleberre do we still need this as part of the setup ?

**Comment by bokobza** [comment]
A few points:
* Is the command `fedsetup` or `federationsetup`?
* The verbs for the commands should not be `Creates`, but `Create` as it's an action that's described.
* `-a` for the multisig could be `-m` instead.
* the creation of multisig is missing the arguments. You need to have a number and some public keys. For example if you want to have a 2 of 3 multisig you need to pass 2 and then 3 pub keys.
* -p the instructions could be clearer here. Also how many are you generating? 1 or 2?

**Comment by ferdeen** [comment]
@bokobza 

> 
> 
> A few points:
> 
>     * Is the command `fedsetup` or `federationsetup`?
> 
>     * The verbs for the commands should not be `Creates`, but `Create` as it's an action that's described.
> 
>     * `-a` for the multisig could be `-m` instead.
> 
>     * the creation of multisig is missing the arguments. You need to have a number and some public keys. For example if you want to have a 2 of 3 multisig you need to pass 2 and then 3 pub keys.
> 
>     * -p the instructions could be clearer here. Also how many are you generating? 1 or 2?

- federationsetup - fedsetup wasn't clear - I've replaced any reference to fedsetup in the code.
- Re verbs okay I've changed
- Changed -a switch to -m
- In master the output usage was `fedkeypairgen [-name=<name>] [-folder=<output_folder>] [-pass=<password>] [-h]` - this wasn't fully implemented.  I'll implement it now.
- The the above is implemented the I'll make the -p switch more clear.

Cheers,

Ferdeen

**Comment by monsieurleberre** [comment]
Maybe somethings to think about:

I have used that package before to handle complex commandline inputs, it helps a lot validating arguments and generating help (up to you):
https://github.com/commandlineparser/commandline

The number of public keys should be enough to know the size of the federation (so no need to pass an N parameter). However it is assumed that the number is odd, and no more than 15 in our spec, maybe a validation for that. Also the quorum for the multisig (M parameter) needs to be >N/2

**Comment by ferdeen** [comment]
Applied changes.  Points of interest :

Run each option in debug.
image

Output for Help.
image

Output for Creating a genesis block.
image

Output for creating public private keys.
image

Output for creating multisig addresses
image

Let me know if we have everything covered. 

Cheers.

**Comment by ferdeen** [comment]
@monsieurleberre just read you point above - let me apply it.  Cheers.

**Comment by bokobza** [comment]
Output for Creating a genesis block: OK
Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?
Output for creating multisig addresses: I see here you're creating many multisig addresses. this was there because this method was used to create all the keys in a test network. 
All you need here is the sidechain multisig address and the Mainchain multisig address. 
image
I also see here that they're the same so maybe there is a problem in the networks used to generate each.

**Comment by ferdeen** [comment]
> 
> 
> Output for Creating a genesis block: OK
> Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?
> Output for creating multisig addresses: I see here you're creating many multisig addresses. this was there because this method was used to create all the keys in a test network.
> All you need here is the sidechain multisig address and the Mainchain multisig address.

Fixed the above. 

With regards to :

> image
> I also see here that they're the same so maybe there is a problem in the networks used to generate each.

@monsieurleberre  I went through this with @bokobza we noticed that the SCRIPT_ADDRESS for all networks are set to 196.  This is why the main and side chain multisig addresses are the same.  I agree with @bokobza it would make sense for these to be different, so it will be easy to identify which multsig address belongs where.  I haven't change the value in the above commit but will probably need to look what [decimal prefix](https://en.bitcoin.it/wiki/List_of_address_prefixes) to use for each network.

image

With regards to :  
> Output for creating public private keys: don't they need to share the public key with which they mine blocks as well?

@monsieurleberre Does the POA feature create a file which contains the federated members public private key for mining ? If so then the current logic I've updated in the console app may be redundant.

**Comment by monsieurleberre** [comment]
as discussed on slack, new Prefixes are coming to avoid confusing networks
also I am not sure about the keys used to mine vs keys used to sign from multisig, I think they could be the same...

**Comment by ferdeen** [comment]
Okay we can check during testing. Cheers.

**Pull request merged**